<a href="https://colab.research.google.com/github/scorpiocodes/Coursera_Capstone/blob/master/Segmenting_and_Clustering_Neighborhoods_in_Toronto_Part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

tutorial link for scrapping a table from a website: https://towardsdatascience.com/web-scraping-html-tables-with-python-c9baba21059

In [0]:
# Import necessary libraries

import requests
import lxml.html as lh
import pandas as pd

In [0]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#Create a handle, page, to handle the contents of the website
page = requests.get(url)

#Store the contents of the website under doc
doc = lh.fromstring(page.content)

#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [3]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

This means that there are 3 columns per row

In [4]:
# Parse the first row as our header
tr_elements = doc.xpath('//tr')

#Create empty list
col=[]
i=0

#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postcode"
2:"Borough"
3:"Neighborhood
"


#### Pandas DataFrame

In [0]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

In [6]:
# Check the length of each column. Ideally, they should all be the same
[len(C) for (title,C) in col]

[287, 287, 287]

This shows that each of the 3 columns has exactly 287 rows

#### Creating the pandas data frame 

In [0]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

In [8]:
# Access the top 5 rows of the data frame 
df.head()

Postcode           Borough      Neighborhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

#### Rearranging and renaming the columns

In [9]:
df.columns = ['Borough', 'Neighbourhood','Postcode']

cols = df.columns.tolist()
cols

cols = cols[-1:] + cols[:-1]

df = df[cols]

df.head()

Postcode Borough     Neighbourhood
0      Not assigned\n     M1A      Not assigned
1      Not assigned\n     M2A      Not assigned
2         Parkwoods\n     M3A        North York
3  Victoria Village\n     M4A        North York
4      Harbourfront\n     M5A  Downtown Toronto

#### Cleaning the messy string in the Borough column

In [10]:
df = df.replace('\n',' ', regex=True)
df.head()

Postcode Borough     Neighbourhood
0      Not assigned      M1A      Not assigned
1      Not assigned      M2A      Not assigned
2         Parkwoods      M3A        North York
3  Victoria Village      M4A        North York
4      Harbourfront      M5A  Downtown Toronto

#### Dropping all cells with a borough that is Not assigned

In [11]:
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

# Reset the index and dropping the previous index
df = df.reset_index(drop=True)

df.head(10)

Postcode Borough     Neighbourhood
0      Not assigned      M1A      Not assigned
1      Not assigned      M2A      Not assigned
2         Parkwoods      M3A        North York
3  Victoria Village      M4A        North York
4      Harbourfront      M5A  Downtown Toronto
5  Lawrence Heights      M6A        North York
6    Lawrence Manor      M6A        North York
7      Queen's Park      M7A  Downtown Toronto
8      Not assigned      M8A      Not assigned
9      Not assigned      M9A      Queen's Park

#### Combining Neighbourhoods based on similar Postcode and Borough  

In [12]:
df = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.columns = ['Postcode','Borough','Neighbourhood']
df.head(10)

Postcode Borough     Neighbourhood
0          Adelaide      M5H  Downtown Toronto
1         Agincourt      M1S       Scarborough
2   Agincourt North      M1V       Scarborough
3    Albion Gardens      M9V         Etobicoke
4         Alderwood      M8W         Etobicoke
5    Bathurst Manor      M3H        North York
6     Bathurst Quay      M5V  Downtown Toronto
7   Bayview Village      M2K        North York
8  Beaumond Heights      M9V         Etobicoke
9      Bedford Park      M5M        North York

#### Removing any space

In [0]:
df['Neighbourhood'] = df['Neighbourhood'].str.strip()

#### Assigning Borough values to the Neignbourhood where vlaue is "Not assigned"

In [0]:
df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

In [15]:
# shape of the data frame
df.shape

(287, 3)

***Saving file to a CSV***

In [0]:
df.to_csv(r'df1.csv')